grpc/src/main/java/com/linecorp/armeria/common/grpc/protocol/AbstractUnaryGrpcService.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.grpc.protocol;
+
+import java.util.concurrent.CompletableFuture;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.grpc.GrpcHeaderNames;
+import com.linecorp.armeria.common.grpc.protocol.ArmeriaMessageDeframer.ByteBufOrStream;
+import com.linecorp.armeria.common.grpc.protocol.ArmeriaMessageDeframer.Listener;
+import com.linecorp.armeria.server.AbstractHttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+
+/**
+ * An {@link AbstractUnaryGrpcService} can be used to implement a gRPC service without depending on gRPC stubs.
+ * This service takes care of deframing and framing with the gRPC wire format and handling appropriate headers.
+ *
+ * <p>This service does not support compression. If you need support for compression, please consider using
+ * normal gRPC stubs or file a feature request.
+ */
+public abstract class AbstractUnaryGrpcService extends AbstractHttpService {
+
+    private static final HttpHeaders RESPONSE_HEADERS =
+            HttpHeaders.of(HttpStatus.OK)
+                       .add(GrpcHeaderNames.GRPC_ENCODING, "identity")
+                       .asImmutable();
+
+    /**
+     * Returns an unframed response message to return to the client, given an unframed request message. It is
+     * expected that the implementation has the logic to know how to parse the request and serialize a response
+     * into {@code byte[]}. The returned {@code byte[]} will be framed and returned to the client.
+     */
+    protected abstract CompletableFuture<byte[]> handleMessage(byte[] message);
+
+    @Override
+    protected final HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req) {
+        final CompletableFuture<HttpResponse> responseFuture =
+                req.aggregateWithPooledObjects(ctx.contextAwareEventLoop(), ctx.alloc())
```

Comment:
Just curious. We have `GrpcServiceBuilder#useBlockingTaskExecutor`, don't we need to respect the flag?

Reply:
This service doesn't use `GrpcServiceBuilder` - it's meant to implement the protocol without gRPC :) As it's a very advanced API, I want to keep its feature set / API surface small so lean towards not adding knobs like that to this class. How does that sound?

Reply:
Ah, I see. This class is definitely not related to the `GrpcServiceBuilder`. I mis-thought about that. Thank you for explanation.